File: apple/Murmur/Home/HomeViewModel.swift
import Foundation
import MurmurShared
import UIKit

/// What the Home screen knows: is Murmur ready, and what did it write last.
///
/// Home is the screen a user lands on every time, so it is also the app's standing health
/// check — a permission revoked in Settings months later shows up here as a chip with a fix,
/// not as a failed dictation.
@MainActor
final class HomeViewModel: ObservableObject {
    @Published var recent: [Transcript] = []
    @Published var mic: PermissionStatus = .notDetermined
    @Published var speech: PermissionStatus = .notDetermined
    /// The engine chip's second line, e.g. "Local · on device" or "Groq · no key".
    @Published var engineSummary: String = ""
    /// `true` when the chosen engine cannot actually run — no on-device model, or a cloud
    /// engine with no key.
    @Published var engineNeedsAttention = false
    /// The row whose "Copied" check is showing.
    @Published var copiedID: Int64?
    /// The Murmur keyboard is in the user's keyboard list.
    @Published var keyboardEnabled = false
    /// The keyboard's Full Access heartbeat — `nil` until the keyboard has appeared once.
    @Published var fullAccess: Bool?

    /// This iPhone has an Action Button, so Home offers to set it up. Hardware never changes
    /// under a running app, so it is read once.
    let hasActionButton: Bool

    /// How many dictations Home shows (spec §5: "the last three dictations").
    static let recentCount = 3

    private let app: AppState
    private let permissions: PermissionsProviding
    private let pasteboard: (String) -> Void
    private let keyboardEnabledProvider: () -> Bool
    private let fullAccessProvider: () -> Bool?
    private var copyResetTask: Task<Void, Never>?

    /// `permissions` and `copy` are injected after `app` so the documented `init(app:)` call
    /// still reads the same, and so the tests never touch the microphone or the pasteboard.
    init(
        app: AppState,
        permissions: PermissionsProviding = LivePermissions(),
        copy: @escaping (String) -> Void = { UIPasteboard.general.string = $0 },
        keyboardEnabledProvider: @escaping () -> Bool = { KeyboardStatus.isEnabled() },
        fullAccessProvider: @escaping () -> Bool? = { KeyboardStatus.hasFullAccess() },
        hasActionButton: Bool = KeyboardStatus.hasActionButton
    ) {
        self.app = app
        self.permissions = permissions
        pasteboard = copy
        self.keyboardEnabledProvider = keyboardEnabledProvider
        self.fullAccessProvider = fullAccessProvider
        self.hasActionButton = hasActionButton
    }

    /// The speech chip only exists while the on-device engine is the one that would run: a
    /// user on Groq has no reason to be nagged about a permission Murmur will never use.
    var showsSpeechChip: Bool {
        app.settings.settings.stt == .local
    }

    var historyUnavailable: Bool { app.historyUnavailable }

    /// Re-reads permissions, the engine choice, the keyboard's two answers and the last three
    /// rows. Called when the screen appears and every time a dictation finishes.
    func reload() {
        mic = permissions.micStatus()
        speech = permissions.speechStatus()

        let stt = app.settings.settings.stt
        let keyPresent: Bool
        if let provider = stt.provider {
            keyPresent = ((try? app.secrets.get(provider.keychainAccount)) ?? nil)?.isEmpty == false
        } else {
            keyPresent = true
        }
        engineSummary = Copy.engineSummary(stt: stt, keyPresent: keyPresent)
        engineNeedsAttention = !keyPresent

        keyboardEnabled = keyboardEnabledProvider()
        fullAccess = fullAccessProvider()

        recent = (try? app.history.recent(Self.recentCount)) ?? []
    }

    func startDictation() {
        app.startInAppDictation()
    }

    /// Same behaviour as a History row: tap copies the cleaned text and shows a check.
    func copy(_ transcript: Transcript) {
        pasteboard(transcript.cleanText)
        copiedID = transcript.id
        copyResetTask?.cancel()
        copyResetTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 1_500_000_000)
            guard !Task.isCancelled else { return }
            await MainActor.run {
                guard let self, self.copiedID == transcript.id else { return }
                self.copiedID = nil
            }
        }
    }

    /// What tapping the microphone chip should do: ask, if the system will still ask; send the
    /// user to Settings once it will not.
    func fixMic() async {
        if mic == .notDetermined {
            _ = await permissions.requestMic()
            mic = permissions.micStatus()
        } else if mic == .denied {
            Permissions.openSettings()
        }
    }

    func fixSpeech() async {
        if speech == .notDetermined {
            _ = await permissions.requestSpeech()
            speech = permissions.speechStatus()
        } else if speech == .denied {
            Permissions.openSettings()
        }
    }
}

File: android/app/src/main/java/com/murmur/app/ui/home/HomeViewModel.kt
package com.murmur.app.ui.home

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import com.murmur.app.data.SecretStore
import com.murmur.app.data.SettingsStore
import com.murmur.app.data.SttEngine
import com.murmur.app.data.history.HistoryStore
import com.murmur.app.data.history.TranscriptEntity
import kotlinx.coroutines.flow.Flow

/**
 * What the Home screen knows: is Murmur ready, and what did it write last.
 *
 * Home is the screen a user lands on every time, so it is also the app's standing health
 * check — a permission revoked in system settings months later shows up here as a chip with a
 * fix, not as a failed dictation. The iOS twin is `HomeViewModel`, minus the chips Android has
 * no equivalent for (speech permission, Full Access, the Action Button).
 */
class HomeViewModel(
    private val settings: SettingsStore,
    private val secrets: SecretStore,
    history: HistoryStore,
    private val micGranted: () -> Boolean,
    private val keyboardEnabled: () -> Boolean,
) : ViewModel() {

    /** The last three dictations, re-emitted the moment the keyboard writes one (spec §5). */
    val recent: Flow<List<TranscriptEntity>> = history.observeRecent(RECENT_COUNT)

    var mic by mutableStateOf(false)
        private set

    /** The dialog has been shown once, so a second denial is worth a trip to app settings. */
    var micAsked by mutableStateOf(false)

    var keyboardOn by mutableStateOf(false)
        private set

    var stt by mutableStateOf(SttEngine.LOCAL)
        private set

    /** Whether the chosen engine has what it needs — always true for the on-device one. */
    var engineReady by mutableStateOf(true)
        private set

    /** The row whose "Copied" check is showing, if any. */
    var copiedId by mutableStateOf<Long?>(null)

    /** Re-reads the permission, the keyboard and the engine. The list flows on its own. */
    suspend fun refresh() {
        mic = micGranted()
        keyboardOn = keyboardEnabled()
        val current = settings.current()
        stt = current.stt
        engineReady = current.stt.provider
            ?.let { !secrets.get(it.keychainAccount).isNullOrEmpty() } ?: true
    }

    fun onMicResult(granted: Boolean) {
        micAsked = true
        mic = granted
    }

    companion object {
        /** How many dictations Home shows (spec §5: "the last three dictations"). */
        const val RECENT_COUNT = 3
    }
}
